File: longest_increasing.java
// Time Complexity :O(nlogn)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int[] greedy = new int[n];
        int length = 1;
        greedy[0] = nums[0];
        // for each element
        for (int i = 1; i < n; i++) {
            if (nums[i] > greedy[length - 1]) {// if current element is greater than last element in greedy array, add
                // in greedy array
                greedy[length] = nums[i];
                length++;
            } else {
                // if not, find elemet just greater or equal to current element using binary
                // search and replace with
                // curr element
                int ind = searchInd(greedy, nums[i], 0, length - 1);
                greedy[ind] = nums[i];
            }

        }
        return length;
    }

    public int searchInd(int[] nums, int target, int start, int end) {
        int low = start;
        int high = end;
        while (low <= high) {
            int mid = low + (end - high) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return low;
    }
}

File: longest_increasing.py
# Time Complexity :
# O(N ^2)

# Space Complexity :
# O(N)

# Did this code successfully run on Leetcode :
#Yes

#We store a DP array that contains the longest increasing subsequence found until that location. 
#We process each element in the array and check if the new element can be added to any of the previous subsequences until then  - if it is (if the current elem is bigger than end of prev subsequence), then current element can be added to the subsequence and we see if this subsequence length is higher than what we already have and update based on that
#In the end, we return the max of DP array - whereever we can find the max longest subsequence

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        
        for i in range(0,n):
            for j in range(0,i):
                if nums[i] > nums[j] :
                    dp[i] = max(dp[j]+1, dp[i])

        return max(dp)
